### PR TITLE
Fix: Detect parsability when FromRoute attribute is used

### LINKED ISSUE #45581: Route parameter TryParse checks does not handle handler parameters with [FromRoute] attribute where name is provided.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Given the following example:

```csharp
using Microsoft.AspNetCore.Mvc;
using System.Diagnostics.CodeAnalysis;

var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.MapGet("/{blah}", ([FromRoute(Name = "blah")]MyRouteObject x) => "Hello World!");

app.Run();

public class MyRouteObject
{
}
```

... the recently added Roslyn analyzer fails to warn that ```MyRouteObject``` does not have a ```TryParse(...)``` method. This results in a runtime error when this code is run.

### Expected Behavior

We should raise an analyzer warning on the handler parameter because it's a route parameter that does not implement ```TryParse```.

### Steps To Reproduce

See above.

### Exceptions (if any)

At runtime we get the following exception:

> InvalidOperationException: No public static bool MyRouteObject.TryParse(string, out MyRouteObject) method found for x.

### .NET Version

8.0.100-alpha.1.22613.3

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by mitchdenny**
…aming.

# Detect parsability even when `FromRoute` attribute is used to map handler parameters to route parameters.

There was a bug in my recent analyzer change where the following wouldn't produce an error:

```csharp
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Mvc;
var webApp = WebApplication.Create();
webApp.MapGet("/customers/{cust}", ([FromRoute(Name = "cust")]Customer customer) => {});

public class Customer {}
```

It should however warn that ```Customer``` type is not parsable.

## Description

To fix I introduced some logic to resolve the route parameter name if the ```FromRoute``` attribute is present. If it is it uses that instead to determine whether to apply the analyzer error. Arguably I could key just off the presence of the ```[FromRoute]``` attribute and not drill into the name, but I think we would want to avoid providing this analyzer error in that case and allow another analyzer to warn on a mismatch between route parameters and handler parameters.

Fixes #45581

### CODE REVIEW

**Comment by JamesNK** [comment]
What happens if you do this?

```cs
webApp.MapGet("/customers/{cust}", ([FromRoute(Name = "cust")]Customer customer, string cust) => {});
```

I'm guessing it explodes at runtime. What happens in the parsability analyzer? (we could write an analyzer that detects this situation and warns about it, but it probably wouldn't get much usage)

**Comment by JamesNK** [comment]
Good find btw. I need to consider using `FromRoute.Name` in some of the tooling I've written recently.

**Comment by mitchdenny** [comment]
> I'm guessing it explodes at runtime.

Oh interesting. It doesn't throw - it actually works. The route parameter value gets put into the string and TryParse gets called. I'm not entirely sure that this should be considered an error. For example the following is also considered valid:

```csharp
app.MapGet("/customers/{cust}", ([FromRoute(Name = "cust")] Customer aaa, [FromRoute(Name = "cust")] Customer customer, string cust) => {});
```

**Comment by JamesNK** [comment]
I guess there isn't anything wrong with multiple arguments binding to the same route value.

**Review by JamesNK — APPROVED** [review]
LGTM

It would be good to add a unit test of multiple arguments binding to one parameter, just to verify that it doesn't break the analyzer. Maybe a comment to the code as well that it doesn't matter if multiple parameters have the same name.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs [hunk 1]
```diff
@@ -86,10 +88,36 @@ public partial class RouteHandlerAnalyzer : DiagnosticAnalyzer
...
+                }
+            }
+
+            return parameterSymbol.Name; // We have the attribute, but name isn't specified!
+        }
+
+        static bool TryGetFromMetadataInterfaceTypeAttributeData(IParameterSymbol parameterSymbol, WellKnownType fromMetadataInterfaceType, WellKnownTypes wellknowTypes, out AttributeData attributeData)
+        {
+            var fromMetadataInterfaceTypeSymbol = wellknowTypes.Get(fromMetadataInterfaceType);
+            attributeData = parameterSymbol.GetAttributes().SingleOrDefault(ad => ad.AttributeClass.Implements(fromMetadataInterfaceTypeSymbol))!;
```
Comment (JamesNK) [inline src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs:114]:
```suggestion
            attributeData = parameterSymbol.GetAttributes().FirstOrDefault(ad => ad.AttributeClass.Implements(fromMetadataInterfaceTypeSymbol))!;
```
It's possible that two attributes implement `IFromRouteMetadata`. Be defensive and use the first one. A unit test with multiple attribute matches would be good as well.

**Comment by JamesNK** [comment]
Fixes in various bits of route tooling for this issue: https://github.com/dotnet/aspnetcore/pull/45720

My PR adds the resolved parameter name to route usage information, so it should be possible to simplify this PR by reusing it.

File: src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs (hunk 1 above)
Comment (JamesNK) [reply src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/DisallowNonParsableComplexTypesOnParameters.cs:114]: My PR resolves this.

**Comment by mitchdenny** [comment]
> It should be possible to simplify this PR by reusing it.

Will pick this up after the holidays and integrate your changes (assuming they are merged by then).

**Comment by mitchdenny** [comment]
@JamesNK I integrated your changes into my PR. It involved moving the ```SymbolExtensions``` type into a more common location since we use it outside of the pattern analyzer stuff now as well. I also hoisted the ```ResolveParameterName(...)``` method into a ```RouteParameterHelper``` class so it can be reused (this is because my logic actually loops over the delegate handler parameters - not the route pattern itself).

Wanted to give you a final chance to review before I squash and merge.

**Comment by JamesNK** [comment]
I made a collection of resolved parameters available from `RouteUsageModel`:

https://github.com/dotnet/aspnetcore/blob/45fcfb75dd7ca6c548200dde4f7374a9a451039b/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/RouteUsageDetector.cs#L25-L44

`RouteParameterName` uses the attribute to update the route name if needed. Could you pull the information you need from it? e.g. loop through the parameter collection, match on the symbol, then use `RouteParameterName`

**Comment by mitchdenny** [comment]
> Could you pull the information you need from it? e.g. loop through the parameter collection, match on the symbol, then use `RouteParameterName`

Are you suggesting something like this:

```csharp
            var parameterSymbol = routeUsage.UsageContext.ResolvedParameters.FirstOrDefault(p => p.Symbol == (ISymbol)handlerDelegateParameter);
            var routeParameterName = parameterSymbol.RouteParameterName ?? handlerDelegateParameter.Name;

            // Match handler parameter against route parameters. If it is a route parameter it needs to be parsable/bindable in some fashion.
            if (routeUsage.RoutePattern.TryGetRouteParameter(routeParameterName, out var routeParameter))
            {
              ...
```

If that is the case, it doesn't seem to work. I can never get ````p.Symbol``` to match ```handlerDelegateParameter``` - which doesn't make sense to me they should be the same object instance right?

**Pull request merged**